Typing failure on expect().toThrow(Error) with explicit imports

<!-- Love Jest? Please consider supporting our collective: 👉  https://opencollective.com/jest/donate -->

## 🐛 Bug Report

Here's the minimum reproducible snippet.

My project is written in Typescript and I'm using `jest` along with `ts-jest`.

```ts
// test.ts
import { test, expect } from '@jest/globals'

test('error', () => {
  function throwError (): never { throw new Error() }
  expect(throwError).toThrow(Error)
})
```

After running the test file, an error is shown.
```
error TS2345: Argument of type 'ErrorConstructor' is not assignable to parameter of type 'string | RegExp | Error | Constructable | undefined'.
      Type 'ErrorConstructor' is not assignable to type 'Constructable'.
        Types of parameters 'message' and 'args' are incompatible.
          Type 'unknown' is not assignable to type 'string | undefined'.
            Type 'unknown' is not assignable to type 'string'.
```

However, if I remove the first line which makes the test using jest globals, everything works fine.

That is,

```ts
// test.ts
test('error', () => {
  function throwError (): never { throw new Error() }
  expect(throwError).toThrow(Error)
})
```

I found that source definition of `expect` is different between two tests.

With explicit imports, the definition is coming from `node_modules\expect\build\types.d.ts`, while with implicit imports, the definition is coming from `node_modules\@types\jest\index.d.ts`.

## To Reproduce

Steps to reproduce the behavior:
- Create a test file with the snippet above.
- Run the test.

## Expected behavior

No typing errors.


## envinfo

```
  System:
    OS: Windows 10 10.0.18362
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  Binaries:
    Node: 12.16.0 - <Program Files>\nodejs\node.EXE
    Yarn: 1.9.4 - <Program Files>\Yarn\bin\yarn.CMD
    npm: 6.13.4 - <Program Files>\nodejs\npm.CMD
  npmPackages:
    jest: ^26.0.1 => 26.0.1
```

This is reproducible with just the `expect` package by itself. Here's how to setup a minimal test:

```sh
$ mkdir repro
$ cd repro
$ npm init
$ npm i --save-dev typescript expect
$ vi test.ts
```

Write the following into the `test.ts` file:

```ts
import expect from 'expect';

function oops() {
    throw new Error('oops');
}

expect(oops).toThrow(Error);
```

Then try to compile it with `--strict --esModuleInterop` (the latter option is needed to allow the use of `import`):

```
$ ./node_modules/.bin/tsc --strict --esModuleInterop test.ts

test.ts:7:22 - error TS2345: Argument of type 'ErrorConstructor' is not assignable to parameter of type 'string | RegExp | Error | Constructable | undefined'.
  Type 'ErrorConstructor' is not assignable to type 'Constructable'.
    Types of parameters 'message' and 'args' are incompatible.
      Type 'unknown' is not assignable to type 'string | undefined'.
        Type 'unknown' is not assignable to type 'string'.

7 expect(oops).toThrow(Error);
                       ~~~~~

  test.ts:7:22
    7 expect(oops).toThrow(Error);
                           ~~~~~
    Did you mean to use 'new' with this expression?


Found 1 error.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 
 - `[expect]` Pass matcher context to asymmetric matchers ([#11926](https://github.com/facebook/jest/pull/11926) & [#11930](https://github.com/facebook/jest/pull/11930))
 - `[expect]` Improve TypeScript types ([#11931](https://github.com/facebook/jest/pull/11931))
+- `[expect]` Improve typings of `toThrow()` and `toThrowError()` matchers ([#11929](https://github.com/facebook/jest/pull/11929))
 - `[@jest/types]` Mark deprecated configuration options as `@deprecated` ([#11913](https://github.com/facebook/jest/pull/11913))
 - `[jest-cli]` Improve `--help` printout by removing defunct `--browser` option ([#11914](https://github.com/facebook/jest/pull/11914))
 - `[jest-haste-map]` Use distinct cache paths for different values of `computeDependencies` ([#11916](https://github.com/facebook/jest/pull/11916))

--- a/packages/expect/src/types.ts
+++ b/packages/expect/src/types.ts
@@ -106,10 +106,6 @@ export type Expect<State extends MatcherState = MatcherState> = {
     not: InverseAsymmetricMatchers & ExtraAsymmetricMatchers;
   };
 
-interface Constructable {
-  new (...args: Array<unknown>): unknown;
-}
-
 // This is a copy from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de6730f4463cba69904698035fafd906a72b9664/types/jest/index.d.ts#L570-L817
 export interface Matchers<R> {
   /**
@@ -322,11 +318,11 @@ export interface Matchers<R> {
   /**
    * Used to test that a function throws when it is called.
    */
-  toThrow(error?: string | Constructable | RegExp | Error): R;
+  toThrow(error?: unknown): R;
   /**
    * If you want to test that a specific error is thrown inside a function.
    */
-  toThrowError(error?: string | Constructable | RegExp | Error): R;
+  toThrowError(error?: unknown): R;
 
   /* TODO: START snapshot matchers are not from `expect`, the types should not be here */
   /**

--- a/test-types/top-level-globals.test.ts
+++ b/test-types/top-level-globals.test.ts
@@ -115,6 +115,12 @@ expectType<void>(describe.skip.each(testTable)(testName, fn, timeout));
 expectType<void>(expect(2).toBe(2));
 expectType<Promise<void>>(expect(2).resolves.toBe(2));
 
+expectType<void>(expect(() => {}).toThrow());
+expectType<void>(expect(() => {}).toThrow(/error/));
+expectType<void>(expect(() => {}).toThrow('error'));
+expectType<void>(expect(() => {}).toThrow(Error));
+expectType<void>(expect(() => {}).toThrow(new Error('error')));
+
 expectType<void>(expect('Hello').toEqual(expect.any(String)));
 
 // this currently does not error due to `[id: string]` in ExtraAsymmetricMatchers - we should have nothing there and force people to use interface merging

